### PR TITLE
For exchanger optimisation

### PR DESCRIPTION
**Pull request by evercodex**
refactoring modal windows and methods for transactions;
several other optimizations

### CODE REVIEW

File: app/includes/emission-content.tpl [hunk 1]
```diff
@@ -80,14 +80,20 @@
       </div>
       <div class="col-sm-11">
         <br/><!-- todo br -> margin/padding? -->
-        <span>Buy rate</span>: {{ buyRate }}</a>
+        <span>Buy rate</span>: {{ buyRate }}
         <br/>
```
Comment (decodedbrain) [inline app/includes/emission-content.tpl]: replace to `<p>`.

File: app/scripts/controllers/bankStatusCtrl.js [hunk 2]
```diff
@@ -113,9 +122,9 @@ var bankStatusCtrl = function($scope, libreService, $translate) {
 
     function fillData(varsObject) {
         return getData(varsObject)
-                .catch((e)=>console.log(e))
-                .then((data)=>processData(data,varsObject))
-                .then((data)=>{
+                .catch((e) => console.log(e))
+                .then((data) => processData(data, varsObject))
+                .then((data) => {
                     console.log(data);
```
Comment (decodedbrain) [inline app/scripts/controllers/bankStatusCtrl.js]: Delete it or check `IS_DEBUG`

File: app/scripts/controllers/emissionCtrl.js [hunk 3]
```diff
@@ -174,22 +174,28 @@ var emissionCtrl = function($scope, $sce, walletService, libreService, $rootScop
...
                     let stateDec = +state.data[0];
                     if ($scope.state != stateDec) {
                         stateWatcher(stateDec);
                     }
                     $scope.state = stateDec;
                     $scope.allowedTokens = allowedTokens.data[0] / Math.pow(10, libreService.coeff.tokenDecimals);
-                    $scope.allTokens = tokenBalance.data[0] / Math.pow(10, libreService.coeff.tokenDecimals);
+                    $scope.allTokens = userTokenBalance.data[0] / Math.pow(10, libreService.coeff.tokenDecimals);
+
+                    $scope.tokenBalance = +exchangerTokenBalance.data[0] / Math.pow(10, libreService.coeff.tokenDecimals);
```
Comment (decodedbrain) [inline app/scripts/controllers/emissionCtrl.js]: Move to number (`+`) convert to data assigment.

File: app/includes/emission-content.tpl [hunk 4]
```diff
@@ -146,9 +157,9 @@
           </label>
         </div>
         <p class="col-xs-12">
-          <span translate="LIBRE_allowed">Allowed: </span><span>{{ allowedTokens | number: 3 }}</span>
+          <span translate="LIBRE_allowed">Allowed:</span> <span>{{ allowedTokens | number: 3 }}</span>
```
Comment (decodedbrain) [inline app/includes/emission-content.tpl]: Maybe delete second span?

File: app/includes/emission-content.tpl [hunk 5]
```diff
@@ -146,9 +157,9 @@
...
           <br/>
-          <span>All tokens: </span><span>{{ allTokens | number: 3 }}</span>
+          <span translate="LIBRE_allTokens">All tokens:</span> <span>{{ allTokens | number: 3 }}</span>
```
Comment (decodedbrain) [inline app/includes/emission-content.tpl]: Maybe delete second span?

File: app/scripts/controllers/emissionCtrl.js [hunk 6]
```diff
@@ -52,33 +52,28 @@ var emissionCtrl = function($scope, $sce, walletService, libreService, $rootScop
...
-    $scope.remissionLimit = gasRemission;
-    $scope.urLimit = gasRUR;
-    // modals
-    $scope.buyModal = new Modal(document.getElementById('buyTx'));
-    $scope.sellModal = new Modal(document.getElementById('sellTx'));
-    $scope.urModal = new Modal(document.getElementById('urTx'));
+
+    $scope.txModal = new Modal(document.getElementById('txModal'));
 
     //$scope.allTokens = 'Loading';
```
Comment (decodedbrain) [inline app/scripts/controllers/emissionCtrl.js]: Delete it if not use

Comment (evercodex) [reply app/scripts/controllers/emissionCtrl.js]: see next commits

File: app/scripts/controllers/emissionCtrl.js [hunk 7]
```diff
@@ -146,8 +140,7 @@ var emissionCtrl = function($scope, $sce, walletService, libreService, $rootScop
         }
     }, 1000);
```
Comment (decodedbrain) [inline app/scripts/controllers/emissionCtrl.js]: Maybe move to const

File: app/scripts/controllers/emissionCtrl.js [hunk 8]
```diff
@@ -167,38 +160,44 @@ var emissionCtrl = function($scope, $sce, walletService, libreService, $rootScop
...
+                        updateRatesCost = values[2],
                         calcTime = values[3],
                         readyOracles = values[4],
                         oracleCount = values[5],
                         requestTime = values[6],
-                        tokenBalance = values[7],
+                        userTokenBalance = values[7],
                         allowedTokens = values[8];
 
+                    ajaxReq.getBalance(bankAddress, function(balanceData) {
```
Comment (decodedbrain) [inline app/scripts/controllers/emissionCtrl.js]: It's may load slowly and will be load later. It's ok?

File: app/scripts/controllers/emissionCtrl.js (hunk 7 above)
Comment (evercodex) [reply app/scripts/controllers/emissionCtrl.js]: 1000 means one second for timer (decreasing values each second)
it shouldn't be changed, so I think we can leave it as is

File: app/scripts/controllers/emissionCtrl.js [hunk 9]
```diff
@@ -294,117 +307,179 @@ var emissionCtrl = function($scope, $sce, walletService, libreService, $rootScop
...
+                        cashAbiRefactor["decreaseApproval"], [bankAddress, allowance - tokensToAllowCount]
+                    );
+                } else {
+                    // tokensToAllowCount > allowance
+                    $scope.tx.data = getDataString(
+                        cashAbiRefactor["increaseApproval"], [bankAddress, tokensToAllowCount - allowance]
+                    );
+                }
+                $scope.tx.to = cashAddress;
+                //$scope.tx.gasLimit = gasApprove;
```
Comment (decodedbrain) [inline app/scripts/controllers/emissionCtrl.js]: Delete it?

File: app/scripts/controllers/emissionCtrl.js [hunk 10]
```diff
@@ -294,117 +307,179 @@ var emissionCtrl = function($scope, $sce, walletService, libreService, $rootScop
...
+                $scope.tx.value = 0;
+                resolve();
```
Comment (decodedbrain) [inline app/scripts/controllers/emissionCtrl.js]: Maybe resolve $scope.tx?

File: app/scripts/controllers/emissionCtrl.js [hunk 11]
```diff
@@ -294,117 +307,179 @@ var emissionCtrl = function($scope, $sce, walletService, libreService, $rootScop
...
-        $scope.tx.value = 0;
-
-        libreTransaction($scope, "sellPending", "SELL", $translate, updateBalanceAndAllowance);
+    function prepareUpdateRatesTx() {
+        return new Promise((resolve, reject) => {
+            getContractData("requestPrice", []).then((oracleDeficit) => {
+                $scope.tx.data = getDataString(bankAbiRefactor["requestRates"], []);
+    
+                $scope.tx.to = bankAddress;
+                //$scope.tx.gasLimit = gasUpdateRates;
```
Comment (decodedbrain) [inline app/scripts/controllers/emissionCtrl.js]: Delete it?

File: app/scripts/controllers/emissionCtrl.js [hunk 12]
```diff
@@ -294,117 +307,179 @@ var emissionCtrl = function($scope, $sce, walletService, libreService, $rootScop
...
+                $scope.tx.value = etherUnits.toEther(+oracleDeficit.data[0], 'wei');
+                $scope.tx.unit = 'ether';
```
Comment (decodedbrain) [inline app/scripts/controllers/emissionCtrl.js]: value in Wei, but unit is ether. it's ok?

File: app/includes/emission-content.tpl (hunk 4 above)
Comment (evercodex) [reply app/includes/emission-content.tpl]: +

File: app/includes/emission-content.tpl (hunk 5 above)
Comment (evercodex) [reply app/includes/emission-content.tpl]: +

File: app/scripts/controllers/emissionCtrl.js (hunk 8 above)
Comment (evercodex) [reply app/scripts/controllers/emissionCtrl.js]: yes

File: app/scripts/controllers/emissionCtrl.js [hunk 13]
```diff
@@ -294,117 +307,179 @@ var emissionCtrl = function($scope, $sce, walletService, libreService, $rootScop
...
-            $scope.tx.data = getDataString(bankAbiRefactor["requestRates"], []);
-
+    $scope.calcRatesTx = function(estimate = false) {
+        if (!estimate) $scope.txModal.close();
+        canCalc().then(function() {
+            $scope.tx.data = getDataString(bankAbiRefactor["calcRates"], []);
             $scope.tx.to = bankAddress;
-            $scope.tx.gasLimit = gasRUR;
-            $scope.tx.value = etherUnits.toEther(+oracleDeficit.data[0], 'wei');
+            //$scope.tx.gasLimit = gasCalcRates;
```
Comment (decodedbrain) [inline app/scripts/controllers/emissionCtrl.js]: Delete it

File: app/scripts/controllers/emissionCtrl.js [hunk 14]
```diff
@@ -294,117 +307,179 @@ var emissionCtrl = function($scope, $sce, walletService, libreService, $rootScop
...
-        canCalc($scope, CRTx);
+    $scope.sellTx = function(estimate = false) {
+        if (!estimate) $scope.txModal.close();
+        canOrder([0, $scope.sellRate]).then(function() {
+            var tokenCount = $scope.tokenValue * Math.pow(10, libreService.coeff.tokenDecimals)
+            $scope.tx.data = getDataString(bankAbiRefactor["sellTokens"], 
+                [$scope.wallet.getAddressString(), tokenCount]);
+    
+            $scope.tx.to = bankAddress;
+            //$scope.tx.gasLimit = gasRemission;
```
Comment (decodedbrain) [inline app/scripts/controllers/emissionCtrl.js]: Delete it

File: app/scripts/controllers/emissionCtrl.js [hunk 15]
```diff
@@ -294,117 +307,179 @@ var emissionCtrl = function($scope, $sce, walletService, libreService, $rootScop
...
+        canOrder([$scope.buyRate, 0]).then(function() {
+            let txData = getDataString(bankAbiRefactor["buyTokens"], [$scope.wallet.getAddressString()]);
 
-        $scope.tx.to = bankAddress;
-        $scope.tx.gasLimit = libreService.coeff.gasCR;
-        $scope.tx.value = 0;
-        $scope.tx.unit = 'ether';
+            $scope.tx.data = txData;
+            $scope.tx.to = bankAddress;
+            //$scope.tx.gasLimit = gasEmission;
```
Comment (decodedbrain) [inline app/scripts/controllers/emissionCtrl.js]: Delete?

File: app/scripts/controllers/emissionCtrl.js (hunk 9 above)
Comment (evercodex) [reply app/scripts/controllers/emissionCtrl.js]: done

File: app/scripts/services/libreService.js [hunk 16]
```diff
@@ -235,15 +260,15 @@ var libreService = function(walletService, $translate) {
                 if (IS_DEBUG) console.log(txData);
                 uiFuncs.generateTx(txData, function(rawTx) {
                     if (rawTx.isError) {
-                        _scope[pendingName] = false;
+                        if (pendingName != null) _scope[pendingName] = false;
                         _scope.notifier.danger("generateTx: " + rawTx.error);
-                        return;
+                        return 0;
```
Comment (decodedbrain) [inline app/scripts/services/libreService.js]: Why zero?

File: app/scripts/controllers/emissionCtrl.js (hunk 10 above)
Comment (evercodex) [reply app/scripts/controllers/emissionCtrl.js]: we don't need any var being resolved from here

File: app/scripts/controllers/emissionCtrl.js (hunk 11 above)
Comment (evercodex) [reply app/scripts/controllers/emissionCtrl.js]: done

File: app/scripts/controllers/emissionCtrl.js (hunk 13 above)
Comment (evercodex) [reply app/scripts/controllers/emissionCtrl.js]: done

File: app/scripts/controllers/emissionCtrl.js (hunk 14 above)
Comment (evercodex) [reply app/scripts/controllers/emissionCtrl.js]: done

File: app/scripts/controllers/emissionCtrl.js (hunk 15 above)
Comment (evercodex) [reply app/scripts/controllers/emissionCtrl.js]: done

File: app/scripts/services/libreService.js [hunk 17]
```diff
@@ -303,118 +325,98 @@ var libreService = function(walletService, $translate) {
...
-
-            if (canOrder)
-                transactionFunc();
-            else {
-                $translate("LIBRE_orderNotAllowed").then((msg) => {
-                    _scope.notifier.danger(msg);
-                });
+    function canOrder(rates = [0, 0]) {
+        return new Promise((resolve, reject) => {
+            if (rates[0] == 0 && rates[1] == 0) {
```
Comment (decodedbrain) [inline app/scripts/services/libreService.js]: What is 0 and 1 in rates? Buy\sell?

File: app/scripts/services/libreService.js (hunk 16 above)
Comment (evercodex) [reply app/scripts/services/libreService.js]: fixed, thx

File: app/scripts/services/libreService.js (hunk 17 above)
Comment (evercodex) [reply app/scripts/services/libreService.js]: yes. for checking buy rate use [buy, 0], for sell - [0, sell]

File: app/scripts/services/libreService.js [hunk 18]
```diff
@@ -303,118 +325,98 @@ var libreService = function(walletService, $translate) {
...
+                $translate("LIBRE_errorValidatingRates").then((msg) => reject(msg));
             }
+            return Promise.all([
+                getContractData("getState"),
+                getContractData(rates[0] != 0 ? "buyRate" : "sellRate")
```
Comment (decodedbrain) [inline app/scripts/services/libreService.js]: Check it

File: app/scripts/services/libreService.js [hunk 19]
```diff
@@ -303,118 +325,98 @@ var libreService = function(walletService, $translate) {
...
+                    resolve();
+                } else {
+                    $translate("LIBRE_RURNotAllowed").then((msg) => reject(msg));
+                }
+            });
         });
     };
 
+    function canCalc() {
+        const MIN_READY_ORACLES = 2;
```
Comment (decodedbrain) [inline app/scripts/services/libreService.js]: Move to service

Comment (evercodex) [reply app/scripts/services/libreService.js]: done

File: app/scripts/services/libreService.js (hunk 18 above)
Comment (evercodex) [reply app/scripts/services/libreService.js]: it's ok

**Pull request closed without merging**